File: node/src/providers/anthropic.ts
/**
 * Anthropic capability ladder — prove depth of access for a leaked API key.
 *
 * A TruffleHog `Anthropic` finding is a secret API key (`sk-ant-...`). The
 * ladder mirrors OpenAI's shape — one safe read-only identity rung, one billable
 * gated rung:
 *
 *   1. `list-models`    `GET /v1/models` — SAFE. The key authenticates and can
 *      list the models available to it (read-only, idempotent, non-billable).
 *      Decides VALID vs DENIED.
 *   2. `create-message` `POST /v1/messages` — GATED. A real message costs the
 *      target money. Wrapped with {@link gated}: the safety boundary runs
 *      *before* any network call, so without BOTH `--prove` and an authorized
 *      scope it throws {@link GatedProbeBlocked} and nothing is sent.
 *
 * The ladder never throws across its public boundary — every failure becomes a
 * {@link ProbeResult}. The raw key is held only transiently and never written
 * into evidence.
 *
 * AUTHORIZED USE ONLY. See TERMS.md.
 */

import { Finding, LadderResult, ProbeResult, Verdict } from "../models.js";
import { redact } from "../redact.js";
import { Consent, GatedProbeBlocked, ProbeTier, gated } from "../safety.js";
import { HttpError, httpRequest, readJson, type FetchLike } from "./http.js";
import { register } from "./registry.js";

/** TruffleHog DetectorName values routed to this ladder (case-insensitive). */
export const DETECTORS = ["Anthropic"] as const;

const API_BASE = "https://api.anthropic.com";
const API_VERSION = "2023-06-01";
const TIMEOUT_MS = 10_000;

function errName(exc: unknown): string {
  if (exc instanceof HttpError) return exc.kind;
  return exc instanceof Error ? exc.constructor.name : "Error";
}

function networkFailure(name: string, tier: ProbeTier, exc: unknown): ProbeResult {
  return new ProbeResult({
    name,
    tier,
    success: false,
    detail: `probe could not complete: ${errName(exc)}`,
    evidence: { error: errName(exc) },
  });
}

function verdictFrom(rungs: ProbeResult[]): Verdict {
  if (rungs.some((r) => r.success && r.tier === ProbeTier.GATED && !r.blocked)) {
    return Verdict.PROVEN;
  }
  if (rungs.some((r) => r.success)) {
    return Verdict.VALID;
  }
  return Verdict.DENIED;
}

/** Anthropic ladder: SAFE `list-models` -> GATED `create-message`. */
export async function anthropicLadder(
  finding: Finding,
  consent: Consent,
  options: { fetchImpl?: FetchLike } = {},
): Promise<LadderResult> {
  const scope = consent.requireLadderScope();
  const rungs: ProbeResult[] = [];
  const key = finding.raw;
  const fetchImpl = options.fetchImpl;

  const models = await anthropicListModels(key, fetchImpl);
  rungs.push(models);

  if (models.success) {
    try {
      rungs.push(await anthropicCreateMessage(consent, key, fetchImpl));
    } catch (exc) {
      if (exc instanceof GatedProbeBlocked) {
        rungs.push(
          new ProbeResult({
            name: "anthropic.create_message",
            tier: ProbeTier.GATED,
            success: false,
            blocked: true,
            detail: `gated billable message blocked: ${exc.reason}`,
            evidence: { reason: exc.reason },
          }),
        );
      } else {
        throw exc;
      }
    }
  }

  return new LadderResult({
    finding,
    provider: "anthropic",
    verdict: verdictFrom(rungs),
    rungs,
    authorizedScope: scope,
  });
}

/** SAFE: `GET /v1/models` confirms the key and lists reachable models. */
async function anthropicListModels(key: string, fetchImpl: FetchLike | undefined): Promise<ProbeResult> {
  const name = "anthropic.list_models";
  let resp: Response;
  try {
    resp = await httpRequest(`${API_BASE}/v1/models`, {
      headers: { "x-api-key": key, "anthropic-version": API_VERSION },
      timeoutMs: TIMEOUT_MS,
      fetchImpl,
    });
  } catch (exc) {
    return networkFailure(name, ProbeTier.SAFE, exc);
  }

  if (resp.status !== 200) {
    return new ProbeResult({
      name,
      tier: ProbeTier.SAFE,
      success: false,
      detail: `key rejected (HTTP ${resp.status})`,
      evidence: { status: resp.status },
    });
  }

  const body = (await readJson(resp)) as { data?: Array<{ id?: string }> } | undefined;
  if (body === undefined) {
    return networkFailure(name, ProbeTier.SAFE, new SyntaxError("invalid JSON"));
  }

  const models = Array.isArray(body.data) ? body.data : [];
  const ids = models.map((m) => m.id).filter((id): id is string => typeof id === "string");
  return new ProbeResult({
    name,
    tier: ProbeTier.SAFE,
    success: true,
    detail: `key authenticates; ${ids.length} models reachable`,
    evidence: {
      status: resp.status,
      key_prefix: redact(key),
      model_count: ids.length,
      sample_models: ids.slice(0, 5),
    },
  });
}

/**
 * GATED: `POST /v1/messages` — a billable message.
 *
 * Wrapped with {@link gated}: the boundary runs before this body, so without
 * BOTH `--prove` and an authorized scope it throws {@link GatedProbeBlocked}
 * and no billable request is ever sent. A minimal `max_tokens` keeps any
 * (consented) spend to the smallest possible amount.
 */
export const anthropicCreateMessage = gated(
  "anthropic.create_message",
  async (_consent: Consent, key: string, fetchImpl?: FetchLike): Promise<ProbeResult> => {
    const name = "anthropic.create_message";
    let resp: Response;
    try {
      resp = await httpRequest(`${API_BASE}/v1/messages`, {
        method: "POST",
        headers: {
          "x-api-key": key,
          "anthropic-version": API_VERSION,
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          model: "claude-3-5-haiku-latest",
          max_tokens: 1,
          messages: [{ role: "user", content: "1" }],
        }),
        timeoutMs: TIMEOUT_MS,
        fetchImpl,
      });
    } catch (exc) {
      return networkFailure(name, ProbeTier.GATED, exc);
    }

    if (resp.status !== 200) {
      return new ProbeResult({
        name,
        tier: ProbeTier.GATED,
        success: false,
        detail: `billable message refused (HTTP ${resp.status})`,
        evidence: { status: resp.status },
      });
    }

    return new ProbeResult({
      name,
      tier: ProbeTier.GATED,
      success: true,
      detail: "billable message creation succeeded (spent the target's credits)",
      evidence: { status: resp.status, billable: true },
    });
  },
);

register([...DETECTORS], (finding, consent) => anthropicLadder(finding, consent));

File: python/src/vtx_recon/providers/anthropic.py
"""Anthropic capability ladder — prove depth of access for a leaked API key.

A TruffleHog ``Anthropic`` finding is a secret API key (``sk-ant-...``). The
ladder mirrors OpenAI's shape — one safe read-only identity rung, one billable
gated rung:

  1. ``list-models``    SAFE. ``GET /v1/models`` — the key authenticates and can
     list the models available to it (read-only, idempotent, non-billable).
     Decides VALID vs DENIED.
  2. ``create-message`` GATED. ``POST /v1/messages`` — a real message costs the
     target money. Wrapped with ``@gated``: the safety boundary runs *before*
     any network call, so without BOTH ``--prove`` and an authorized scope it
     raises ``GatedProbeBlocked`` and nothing is sent.

The ladder never raises across its public boundary — every failure becomes a
:class:`ProbeResult`. The raw key is held only transiently and never written
into evidence.

AUTHORIZED USE ONLY. See TERMS.md.
"""

from __future__ import annotations

import httpx

from ..models import Finding, LadderResult, ProbeResult, Verdict
from ..redact import redact
from ..safety import Consent, GatedProbeBlocked, ProbeTier, gated
from . import register

__all__ = ["DETECTORS", "anthropic_ladder"]

# TruffleHog DetectorName values routed to this ladder (matched case-insensitively).
DETECTORS = ("Anthropic",)

_API_BASE = "https://api.anthropic.com"
_API_VERSION = "2023-06-01"
_TIMEOUT = httpx.Timeout(10.0)


def _network_failure(name: str, tier: ProbeTier, exc: Exception) -> ProbeResult:
    """Turn an httpx/transport error into a non-success rung (never raise)."""
    return ProbeResult(
        name=name,
        tier=tier,
        success=False,
        detail=f"probe could not complete: {type(exc).__name__}",
        evidence={"error": type(exc).__name__},
    )


def _verdict_from(rungs: list[ProbeResult]) -> Verdict:
    """Derive the impact tier from the rungs that ran."""
    if any(r.success and r.tier is ProbeTier.GATED and not r.blocked for r in rungs):
        return Verdict.PROVEN
    if any(r.success for r in rungs):
        return Verdict.VALID
    return Verdict.DENIED


@register("Anthropic")
async def anthropic_ladder(finding: Finding, consent: Consent) -> LadderResult:
    """Anthropic ladder: SAFE ``list-models`` -> GATED ``create-message``."""
    scope = consent.require_ladder_scope()
    rungs: list[ProbeResult] = []
    key = finding.raw

    models = await _anthropic_list_models(key)
    rungs.append(models)

    if models.success:
        try:
            rungs.append(await _anthropic_create_message(consent, key))
        except GatedProbeBlocked as blocked:
            rungs.append(
                ProbeResult(
                    name="anthropic.create_message",
                    tier=ProbeTier.GATED,
                    success=False,
                    blocked=True,
                    detail=f"gated billable message blocked: {blocked.reason}",
                    evidence={"reason": blocked.reason},
                )
            )

    return LadderResult(
        finding=finding,
        provider="anthropic",
        verdict=_verdict_from(rungs),
        rungs=rungs,
        authorized_scope=scope,
    )


async def _anthropic_list_models(key: str) -> ProbeResult:
    """SAFE: ``GET /v1/models`` confirms the key and lists reachable models."""
    name = "anthropic.list_models"
    try:
        async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
            resp = await client.get(
                f"{_API_BASE}/v1/models",
                headers={"x-api-key": key, "anthropic-version": _API_VERSION},
            )
    except httpx.HTTPError as exc:
        return _network_failure(name, ProbeTier.SAFE, exc)

    if resp.status_code != 200:
        return ProbeResult(
            name=name,
            tier=ProbeTier.SAFE,
            success=False,
            detail=f"key rejected (HTTP {resp.status_code})",
            evidence={"status": resp.status_code},
        )

    try:
        body = resp.json()
    except ValueError as exc:
        return _network_failure(name, ProbeTier.SAFE, exc)

    data = body.get("data") if isinstance(body, dict) else None
    models = data if isinstance(data, list) else []
    ids = [m.get("id") for m in models if isinstance(m, dict) and isinstance(m.get("id"), str)]
    return ProbeResult(
        name=name,
        tier=ProbeTier.SAFE,
        success=True,
        detail=f"key authenticates; {len(ids)} models reachable",
        evidence={
            "status": resp.status_code,
            "key_prefix": redact(key),
            "model_count": len(ids),
            "sample_models": ids[:5],
        },
    )


@gated("anthropic.create_message")
async def _anthropic_create_message(_consent: Consent, key: str) -> ProbeResult:
    """GATED: ``POST /v1/messages`` — a billable message.

    Decorated with ``@gated``: the boundary runs before this body, so without
    BOTH ``--prove`` and an authorized scope it raises ``GatedProbeBlocked`` and
    no billable request is ever sent. A minimal ``max_tokens`` keeps any
    (consented) spend to the smallest possible amount.
    """
    name = "anthropic.create_message"
    try:
        async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
            resp = await client.post(
                f"{_API_BASE}/v1/messages",
                headers={
                    "x-api-key": key,
                    "anthropic-version": _API_VERSION,
                    "Content-Type": "application/json",
                },
                json={
                    "model": "claude-3-5-haiku-latest",
                    "max_tokens": 1,
                    "messages": [{"role": "user", "content": "1"}],
                },
            )
    except httpx.HTTPError as exc:
        return _network_failure(name, ProbeTier.GATED, exc)

    if resp.status_code != 200:
        return ProbeResult(
            name=name,
            tier=ProbeTier.GATED,
            success=False,
            detail=f"billable message refused (HTTP {resp.status_code})",
            evidence={"status": resp.status_code},
        )

    return ProbeResult(
        name=name,
        tier=ProbeTier.GATED,
        success=True,
        detail="billable message creation succeeded (spent the target's credits)",
        evidence={"status": resp.status_code, "billable": True},
    )
